File: server/src/routes/hierarchy.ts
import { Router } from 'express';
import { asyncHandler } from '../middleware/error.js';
import { readHierarchy, writeHierarchy } from '../hierarchy/store.js';
import type { HierarchyMap } from '../hierarchy/store.js';

const router = Router();

/** GET /api/hierarchy — returns the current hierarchy map. */
router.get(
  '/',
  asyncHandler(async (_req, res) => {
    const hierarchy = await readHierarchy();
    res.json(hierarchy);
  }),
);

/** PUT /api/hierarchy — replaces the entire hierarchy map. */
router.put(
  '/',
  asyncHandler(async (req, res) => {
    const body = req.body;

    if (typeof body !== 'object' || body === null || Array.isArray(body)) {
      res.status(400).json({ error: 'Invalid hierarchy' });
      return;
    }

    const saved = await writeHierarchy(body as HierarchyMap);
    res.json(saved);
  }),
);

export default router;

File: server_py/app/routes/hierarchy.py
from fastapi import APIRouter, Body, HTTPException

from app.hierarchy.store import HierarchyMap, read_hierarchy, write_hierarchy


router = APIRouter(prefix="/api/hierarchy")


def _is_valid_hierarchy(value: object) -> bool:
    if not isinstance(value, dict):
        return False
    return all(isinstance(item, str) or item is None for item in value.values())


@router.get("")
async def get_hierarchy() -> HierarchyMap:
    return await read_hierarchy()


@router.put("")
async def put_hierarchy(body: object = Body(...)) -> HierarchyMap:
    if not _is_valid_hierarchy(body):
        raise HTTPException(status_code=400, detail={"error": "Invalid hierarchy"})
    return await write_hierarchy(body)
